Adding tigerfacilityservices.com to showcase

flaviolivolsi opened the pull request.

jlengstorf commented:
Hey, @flaviolivolsi — would you be willing to share your i18n setup? We're looking to write up docs and/or build some tooling to make i18n easier, so if you're willing to share your setup it could give us a head start. cc @gatsbyjs/gatsby-admin 

Thanks!

flaviolivolsi commented:
Hey @jlengstorf, I'm glad to!

Our i18n is handled by `react-intl` and the Contentful built-in localization (using of course `gatsby-source-contentful`). Contentful is used for content that changes from page to page while being based on the same shared template. When instead we need something more static (e.g.: the homepage for example has a specific unique template) we use translation strings with `react-intl`.

Follows an overview of how the code is structured.

---

`src/data/languages.js`: here there are specified the enabled languages for each country, it's needed at build time to build each page:

```
module.exports = {
    country_langs: {
        de: ['en', 'de'],
        ch: ['en', 'de', 'fr']
    }
};
```

---

`gatsby-node.js`: pages are built - most of them Contentful and a few manually - in a way that enables us to localize also the URL (an example is the Imprint page, there are both [English](https://www.tigerfacilityservices.com/de-en/imprint/) and [German](https://www.tigerfacilityservices.com/de-de/impressum/) versions and they belong to the same Contentful entry):

```
const languages = require(`./src/data/languages`);

exports.createPages = ({ graphql, boundActionCreators }) => {
    ...
    return new Promise((resolve, reject) => {
        graphql(
            `
                {
                    allContentfulSeoPage(limit: 1000) {
                        edges {
                            node {
                                id
                                contentfulid
                                node_locale
                                country
                                path
                                forceSingleLanguage
                            }
                        }
                    }
                }
            `
        )
        .then(result => {
            if (result.errors) {
                reject(result.errors);
            }

            _.each(result.data.allContentfulSeoPage.edges, edge => {
                if (languages.country_langs[edge.node.country].indexOf(edge.node.node_locale) === -1) {
                    return;
                }

                if (edge.node.forceSingleLanguage && edge.node.node_locale !== edge.node.forceSingleLanguage) {
                    return;
                }

                createPage({
                    path: `/${edge.node.country}-${edge.node.node_locale}/${edge.node.path}/`,
                    component: slash(path.resolve(`./src/templates/seo_page/index.js`)),
                    context: {
                        id: edge.node.id,
                        contentfulid: edge.node.contentfulid,
                        country: edge.node.country,
                        language: edge.node.node_locale,
                    },
                });
            });
        })
    });
    ...
};
```

---

`intl.js`: provides an util to get the requested locale:

```
import config             from "./config";

export const current_locale = (pathname) => {
    let locale;

    if (pathname) {
        locale = pathname.split("/")[1];
    } else {
        // -- We check whether window is undefined because of a Webpack error
        locale = typeof window === "undefined" ? "de-en" : window.location.pathname.split("/")[1];
    }

    const language = locale.split("-")[1];

    if (config.locales.includes(locale)) {
        return locale;
    }

    return "de-en";
};
```

---

`src/layouts/index.js`: wraps `children()` in the `IntlProvider` which loads the locales from the namesake directory (the content is `de-de.json`, `de-en.json` and so on):

```
import "babel-polyfill";
import "intl";
import "intl/locale-data/jsonp/en.js";
import React                from "react";
import { IntlProvider }     from "react-intl";
import { current_locale }   from "../intl";
import locales              from "../locales";
import utils                from "../utils";
import                           "../main.scss";

class TemplateWrapper extends React.Component {
    render() {
        return (
            <IntlProvider
                locale={current_locale(this.props.location.pathname)}
                messages={locales[current_locale(this.props.location.pathname)]}>
                <main className="main">
                    {this.props.children()}
                </main>
            </IntlProvider>
        );
    }
}

export default TemplateWrapper;
```

---

And that's about it. If I can be of more help, please let me know!

araujoigor commented:
@flaviolivolsi very nice! I would only add that one of the best things of our setup is the power that gatsby bring us when combined with the contentful internationalization capabilities:

Our queries return nodes for different languages, with different contents but that can be related to each other by the `contentfulid` (so you know that one page is the translation of another one because of the `contentfulid`. Therefore, after we pass this data to our language selector component it knows which languages/translations the current page has and what is the URL for each one of them.

We will write a blog post about the setup in the future and would be happy to share or to use it to help set up a documentation :)
